### ISSUE TITLE
Default XGBoost binary:logistic objective vs self-implemented objective yields different results

### ISSUE THREAD

**Issue by selectasterisk** [dmlc/xgboost#3892]
Hi, I'm using the python implementation of XGBoost for a classification task. I recently started trying out different loss functions, and noticed that by plugging in the vanilla log loss function found here:
https://github.com/dmlc/xgboost/blob/master/demo/guide-python/custom_objective.py#L20

that I was getting completely different results vs using 'binary:logistic'.

I was wondering if anyone noticed something similar or if someone could tell me what is wrong?

**Comment by hcho3** [comment]
Take a look at the comment:
https://github.com/dmlc/xgboost/blob/0ddb8a7661013c75674704872c22273415268dbd/demo/guide-python/custom_objective.py#L29-L33

**Comment by hcho3** [comment]
Yes

**Comment by selectasterisk** [comment]
So given that's the case, and that predict gives proper transformed outputs, is there another reason that the results are so drastically different? AUC and predicted distributions look completely different, and they are derived from bst.predict.

**Comment by hcho3** [comment]
I think you should define a custom evaluation function to make a fair comparison. Can you post your script so that I can look at it?

**Comment by selectasterisk** [comment]
`
def logregobj(preds, dtrain):
    labels = dtrain.get_label()
    preds = 1.0 / (1.0 + np.exp(-preds))
    grad = preds - labels
    hess = preds * (1.0 - preds)
    return grad, hess

X_train, X_test, y_train, y_test = train_test_split(data_sparse_rw, labels_rw, test_size=0.33, random_state=42)

dtrain = xgb.DMatrix(data=X_train,label=y_train)

dtest = xgb.DMatrix(data=X_test,label=y_test)



num_round = 150
test1_xgb = xgb.train(param, dtrain, num_round
                      ,obj=logregobj
                     )
# make prediction
test1_xgb_probs = test1_xgb.predict(dtest)
roc_auc_score(y_test, test1_xgb_probs)
`


commenting and uncommenting the customized objective function gives clearly different results (with my dataset at least)

**Comment by hcho3** [comment]
@selectasterisk You are using `scale_pos_weight`, which needs to be applied to the custom objective function too:
```python
scale_pos_weight = 3.0

def logregobj(preds, dtrain):
    labels = dtrain.get_label()
    weights = np.where(labels == 1.0, scale_pos_weight, 1.0)
    #preds = 1.0 / (1.0 + np.exp(-preds))
    grad = preds - labels
    hess = preds * (1.0 - preds)
    return grad * weights, hess * weights
```

Also, since `'objective` is set to `binary:logistic`, the `preds` value is already transformed probability, so we should comment out the line `preds = 1.0 / (1.0 + np.exp(-preds))`. With the two changes, the custom objective produces the same result as `binary:logistic`.
